R/ark.R

```diff
@@ -57,39 +64,87 @@ list_tables <- function(db){
   DBI::dbListTables(db)
 }
 
-#' @importFrom DBI dbSendQuery dbFetch
+#' @importFrom DBI dbSendQuery dbFetch dbClearResult dbGetQuery
 ark_file <- function(tablename, 
                      db_con, 
                      lines = 10000L, 
                      dir = ".", 
-                     compress = c("bzip2", "gzip", "xz", "none")){
+                     compress = c("bzip2", "gzip", "xz", "none"),
+                     use_alternate = FALSE){
   
+  ## Set up compressed connection
   compress <- match.arg(compress)
-  size <- DBI::dbGetQuery(db_con, paste("SELECT COUNT(*) FROM", tablename))
-  end <- size[[1]][[1]]
+  ext <- switch(compress,
+                "bzip2" = ".bz2",
+                "gzip" = ".gz",
+                "xz" = ".xz",
+                "none" = "",
+                ".bz2")
+  filename <- file.path(dir, paste0(tablename, ".tsv", ext))
+  con <- compressed_file(filename, "w")
+  on.exit(close(con))
   
-  start <- 1
+  ## Progress reporting
+  message(sprintf("Exporting %s in %d line chunks:", tablename, lines))
   p <- progress::progress_bar$new("[:spin] chunk :current", total = 100000)
-  message(sprintf("Exporting %s in %d line chunks:",
-                  tablename, lines))
   t0 <- Sys.time()
+ 
+  if(use_alternate){
+    alternate_method(db_con, lines, dir, compress, p, tablename, con)
+  } else {
+    kirill_method(db_con, lines, p, tablename, con)
+  }
+  
+  
+  
+  message(sprintf("\t...Done! (in %s)", format(Sys.time() - t0)))
+  
+}
+
+
+kirill_method <- function(db_con, lines, p, tablename, con){
+  ## Create header to avoid duplicate column names
+  query <- paste("SELECT * FROM", tablename, "LIMIT 0")
+  header <- DBI::dbGetQuery(db_con, query)
+  readr::write_tsv(header, con, append = FALSE)
```

Comment:
I wonder if we can get rid of this `dbGetQuery()` call and handle also the initialization in the loop.

Reply:
That would be great.  Any suggestions?  (I tried just commenting this out, which kinda works with the `readr` functions since they actually don't mind the headers being repeated, `readr::read_tsv()` detects this and remove it, but clearly not a robust solution).